syntax = "proto3";

package ark_msgs;

// Container for all messages sent over Ark network
message Envelope {

  enum EndpointType {
    UNSPECIFIED = 0;
    PUBLISH = 1;
    RECEIVE = 2;
    REQUEST = 3;
    RESPONSE = 4;
  }

  EndpointType endpoint_type = 1;
  string channel = 2;
  string src_node_name = 3;
  string dst_node_name = 4; // not set for PUBLISH/REQUEST type
  uint64 sent_timestamp = 5;
  uint64 recv_timestamp = 6;
  string msg_type = 7;
  bytes payload = 8;
  Envelope req_msg = 9; // only used in RESPONSE types
}
